### PR TITLE
Fix subprocess encoding on (non-english) windows and issue 21373, 21374

### PR DESCRIPTION
**Pull request by rkanavath**
## Description

fixes encoding issue (non-english) on Windows and # 21373 # 21374 

Regarding windows encoding, a fix was already there for GRASS and we copied it to OTB. Other providers may also be affected on a non-english windows ?

Regarding issue # 21373
Handle optional status of parameters based on user-input.

Parameters are marked required in descriptor file for certian applications but they are only required if a parent parameter has a "certain" value.  `OtbParameterChoice` and its wrapper will update optional status of all sub-parameters depending on the value of a choice parameter.

Here is an example based on Segmentation application from which issue was spotted. (only relevant lines from descriptor files are listed)

```
Segmentation|mode.raster.out
QgsProcessingParameterRasterLayer|in|Input Image|None|False
OTBParameterChoice|filter|Segmentation algorithm|meanshift;cc;watershed;mprofiles|0|True
QgsProcessingParameterNumber|filter.meanshift.spatialr|Spatial radius|QgsProcessingParameterNumber.Integer|5|True
QgsProcessingParameterNumber|filter.meanshift.ranger|Range radius|QgsProcessingParameterNumber.Double|15|True
QgsProcessingParameterString|filter.cc.expr|Condition|None|False|False
QgsProcessingParameterNumber|filter.watershed.threshold|Depth Threshold|QgsProcessingParameterNumber.Double|0.01|True
QgsProcessingParameterNumber|filter.mprofiles.size|Profile Size|QgsProcessingParameterNumber.Integer|5|True
OTBParameterChoice|mode|Processing mode|vector;raster|0|True
QgsProcessingParameterFileDestination|mode.vector.out|Output vector file
OTBParameterChoice|mode.vector.outmode|Writing mode for the output vector file|ulco;ovw;ulovw;ulu|0|True
QgsProcessingParameterRasterLayer|mode.vector.inmask|Mask Image|None|True
QgsProcessingParameterString|mode.vector.ogroptions|OGR options for layer creation|None|True|True
QgsProcessingParameterRasterDestination|mode.raster.out|Output labeled image
*QgsProcessingParameterEnum|outputpixeltype|Output pixel type|uint8;int;float;double|False|2|True
```
In the above example 
`QgsProcessingParameterString|filter.cc.expr|Condition|None|False|False`
parameter 'filter.cc.expr' is required only if value of 'filter' parameter is 'cc'. So it cannot be made optional in descriptor file.

So depending on value of 'filter' (type `OtbParameterChoice`) some of "sub-parameters" can stay optional or turn to be required. This  bug is fixed in OtbParameterChoice class.

Regarding issue # 21374 

Processing provider checks for output parameter value using `isSupportedOutputValue()`. This method is called in `AlgorithmDialog` class and on failure it raises an exception.
For output parameters that are optional, it is not interesting to raise an exception because they are not created.

## Checklist

> Reviewing is a process done by project maintainers, mostly on a volunteer basis. We try to keep the overhead as small as possible and appreciate if you help us to do so by completing the following items. Feel free to ask in a comment if you have troubles with any of them.

- [x] Commit messages are descriptive and explain the rationale for changes
- [x] Commits which fix bugs include `fixes #11111` in the commit message next to the description
- [x] I have read the [QGIS Coding Standards](https://docs.qgis.org/testing/en/docs/developers_guide/codingstandards.html) and this PR complies with them
- [x] This PR passes all existing unit tests (test results will be reported by travis-ci after opening this PR)
- [x] New unit tests have been added for core changes
- [x] I have run [the `scripts/prepare-commit.sh` script](https://github.com/qgis/QGIS/blob/master/.github/CONTRIBUTING.md#contributing-to-qgis) before each commit

### CODE REVIEW

File: python/plugins/processing/gui/AlgorithmDialog.py [hunk 1]
```diff
@@ -165,9 +165,11 @@ def getParameterValues(self):
                     parameters[param.name()] = value
                     if param.isDestination():
                         context = dataobjects.createContext()
+                        #TODO: could we ignore below check if parameter is optional?
```
Comment (nyalldawson) [inline python/plugins/processing/gui/AlgorithmDialog.py]: Yes - looks like a bug in isSupportedOutputValue. Can you drop this commit from the PR and I'll fix the c++ method?

Comment (nyalldawson) [reply python/plugins/processing/gui/AlgorithmDialog.py]: Fixed in https://github.com/qgis/QGIS/pull/9554

Comment (rkanavath) [reply python/plugins/processing/gui/AlgorithmDialog.py]: Cool!.
 I will update this PR

Comment (nyalldawson) [reply python/plugins/processing/gui/AlgorithmDialog.py]: Ok, #9554 is merged now

**Pull request merged**